### DISCUSSION TITLE
Create a goal conditioned wrapper for the environments.

### DISCUSSION THREAD

**Discussion by virajparimi** [PKU-Alignment/safety-gymnasium#115] [Q&A]
Hi,

I have been using the Safe Point Goal environment. However, I need to write a goal-conditioned wrapper around it for my work. I have noticed that there is a `sample_goal_position` function in the `random_generator` of the task of the environment but when i use it to sample a valid goal position it complains the following,

`in sample_goal_position
    placements, keepout = self.placements['goal']
TypeError: 'NoneType' object is not subscriptable`

I have noticed that the `placements_info.placements` is indeed empty so I need to initialize it. However I am not sure which API to use for that as I found the `build_placements_dict` function that is abstract. 

Can you point me to some documentation or help me out with pointers on how can I create a goal-conditioned wrapper for the environments supported by safety-gymnasium?

**Comment by muchvo** [comment]
The specific implementation of the `_build_placements_dict` function is found in [this file](https://github.com/PKU-Alignment/safety-gymnasium/blob/main/safety_gymnasium/bases/base_task.py#L276-L277). Additionally, the sampling of the goal position information occurs within [this function](https://github.com/PKU-Alignment/safety-gymnasium/blob/main/safety_gymnasium/bases/base_task.py#L352-L353). Could you please provide more details about what the goal-conditioned wrapper needs to accomplish? This will help me offer more direct assistance.

**Comment by virajparimi** [comment]
Hi,

I was able to get my custom goal-conditioned wrapper implemented. The issue was that those attributes are only populated after the reset and I was not doing that before. However, I have a followup question. Right now I use the `make` function to generate the environment with the string ID. However, This generates an environment with bunch of prebuilt wrappers around the environment. For example it already includes the TimeLimit wrapper. I want to include my wrapper before the TimeLimit wrapper is added around the environment. Is there a way to this?

**Comment by muchvo** [comment]
1. Do you mean that you would like to customize the placement of your goal through a wrapper?

2. Yes, that's possible. Specifically, you'll need to reference the `make` [function here](https://github.com/PKU-Alignment/safety-gymnasium/blob/main/safety_gymnasium/utils/registration.py#L207-L208). You just need to modify this function to add your wrapper. Alternatively, you could remove the automatic wrapping of the `TimeLimit` wrapper and define the order yourself:

   ```python
   # before
   __register_helper(
       env_id='SafetyHalfCheetahVelocity-v1',
       entry_point='safety_gymnasium.tasks.safe_velocity.safety_half_cheetah_velocity_v1:SafetyHalfCheetahVelocityEnv',
       max_episode_steps=1000,
       reward_threshold=4800.0,
   )

   # after
   __register_helper(
       env_id='SafetyHalfCheetahVelocity-v1',
       entry_point='safety_gymnasium.tasks.safe_velocity.safety_half_cheetah_velocity_v1:SafetyHalfCheetahVelocityEnv',
       reward_threshold=4800.0,
   )
   env = safety_gymnasium.make('SafetyHalfCheetahVelocity-v1')
   env = YourWrapper(env)
   env = SafeTimeLimit(env, 1000)
   ```
Additionally, you can utilize the `additional_wrappers` parameter to define the order of wrappers, both before and after they are applied. For more specific guidance, please refer to [this documentation](https://gymnasium.farama.org/api/registry/). Note that this approach might require more extensive configuration.
